Make Highway timers configurable; request state periodically from a random peer.

afck opened the pull request:
This separates the standstill alert — restart if no progress is made — from requesting the latest protocol state, and makes those timers configurable separately. The latest protocol state is requested from only one peer instead of from everyone.

Closes #1914.

afck committed:
Request latest state only from random peers, not everyone.

afck committed:
Make Highway timers configurable.

afck committed:
Changelog

afck committed:
Silence Clippy warning in a test.

goral09 commented on node/src/components/consensus/protocols/highway/tests.rs:
```diff
@@ -95,13 +94,14 @@ where
         0,
         start_timestamp,
     );
-    // We expect for messages:
```
😅

goral09 commented on utils/nctl/sh/scenarios/configs/gov96.config.toml:
```diff
@@ -47,16 +47,18 @@ unit_hashes_folder = "../node-storage"
-# from peers.
-standstill_timeout = '30sec'
+# If the current era's protocol state has not progressed for this long, shut down.
+# standstill_timeout = '30min'
```
Why are commenting it out? That could in theory affect the test outcome.

afck replied:
Good point: I guess it's best to keep it enabled and set to something shorter (3 minutes?) in the tests, so they don't just keep running if they go wrong.

goral09 replied:
A good start would be to simply leave it as it was before - but we have to remember that it actually was twice that much b/c it used that timer twice before shutting down.

afck replied:
You're right, there's no reason to change that behavior in this PR; I reset them.

goral09 commented on node/src/components/consensus/protocols/highway.rs:
```diff
@@ -471,76 +472,56 @@ impl<I: NodeIdT, C: Context + 'static> HighwayProtocol<I, C> {
-            now + self.standstill_timeout,
-            TIMER_ID_STANDSTILL_ALERT,
-        )]
+        // We haven't made any progress. Request latest panorama from peers and schedule
```
Aren't we requesting the latest state regardless of whether we made any progress or not?

goral09 reviewed (approved):
Looks good. I'm wondering whether we should base the state request timer on the round length. Making a request every 5/10/20s could be too much if rounds can be at fastest 60s etc. Although, that doesn't work the other way around (I think): if rounds are really short then we shouldn't need to make the request that often since we should keep receiving units more often (shorter rounds).

OK, I think I've convinced myself that we don't have to base that timer on the round length.

afck committed:
Fix comment.

afck committed:
Enable standstill timeout for tests.

In tests we want to be alerted if something goes wrong and a node
doesn't make progress.

afck commented:
@goral09: Thanks, I addressed your two comments. I'm not sure it's a good idea to apply some formula to determine the request frequency: Giving the interval explicitly in the config is easier to understand. And the overhead should be minimal anyway: The request consists of one panorama, so it's about the size of a unit, and we send it to one peer at a time. In comparison, we send two units to _every_ peer in every round.

goral09 commented:
I agree @afck 👍

afck committed:
Don't change standstill timeout values in tests.

Reset them back to what they were — or rather twice that value, because
that was the actual time until restart in the old code.

fizyk20 commented:
bors merge
